File: matrix-core/src/main/java/io/matrix/federation/liquid/FederationMap.java
package io.matrix.federation.liquid;

import java.util.*;

/**
 * W588 — Federation Map.
 *
 * Force-directed graph of nodes, roles, and consensus links.
 * Generates visualization data for browser rendering.
 */
public final class FederationMap {

    /**
     * A node in the federation map.
     */
    public record MapNode(
            long id,
            String label,
            NodeRole role,
            double x,
            double y,
            Map<String, String> metadata
    ) {}

    /**
     * An edge in the federation map.
     */
    public record MapEdge(
            long fromId,
            long toId,
            String type,
            double weight
    ) {}

    private final Map<Long, MapNode> nodes = new HashMap<>();
    private final List<MapEdge> edges = new ArrayList<>();
    private final Random rng;

    public FederationMap(long seed) {
        this.rng = new Random(seed);
    }

    /**
     * Add a node to the map.
     */
    public void addNode(long id, String label, NodeRole role) {
        double x = rng.nextDouble() * 800;
        double y = rng.nextDouble() * 600;
        nodes.put(id, new MapNode(id, label, role, x, y, Map.of()));
    }

    /**
     * Add an edge between nodes.
     */
    public void addEdge(long fromId, long toId, String type, double weight) {
        edges.add(new MapEdge(fromId, toId, type, weight));
    }

    /**
     * Generate JSON for visualization.
     */
    public String toJson() {
        StringBuilder sb = new StringBuilder("{");

        // Nodes
        sb.append("\"nodes\":[");
        boolean first = true;
        for (MapNode node : nodes.values()) {
            if (!first) sb.append(",");
            sb.append("{\"id\":").append(node.id())
              .append(",\"label\":\"").append(node.label()).append("\"")
              .append(",\"role\":\"").append(node.role()).append("\"")
              .append(",\"x\":").append(String.format("%.1f", node.x()))
              .append(",\"y\":").append(String.format("%.1f", node.y()))
              .append(",\"color\":\"").append(roleToColor(node.role())).append("\"")
              .append("}");
            first = false;
        }
        sb.append("],");

        // Edges
        sb.append("\"edges\":[");
        first = true;
        for (MapEdge edge : edges) {
            if (!first) sb.append(",");
            sb.append("{\"from\":").append(edge.fromId())
              .append(",\"to\":").append(edge.toId())
              .append(",\"type\":\"").append(edge.type()).append("\"")
              .append(",\"weight\":").append(edge.weight())
              .append("}");
            first = false;
        }
        sb.append("]}");

        return sb.toString();
    }

    /**
     * Generate HTML visualization.
     */
    public String toHtml() {
        return "<!DOCTYPE html><html><head><title>MATRIX Federation Map</title>" +
            "<script src='https://d3js.org/d3.v7.min.js'></script>" +
            "<style>body{font-family:monospace;background:#0a0a0a;padding:20px;}" +
            "h1{color:#00ffff;}svg{background:#111;border:1px solid #333;}" +
            ".node{stroke:#00ff88;stroke-width:2px;}" +
            ".link{stroke:#333;stroke-width:1px;}" +
            ".label{fill:#00ff88;font-size:12px;}</style></head><body>" +
            "<h1>MATRIX Federation Map</h1>" +
            "<svg id='map' width='800' height='600'></svg>" +
            "<script>" +
            "const data=" + toJson() + ";" +
            "const svg=d3.select('#map');" +
            "svg.selectAll('line').data(data.edges).enter().append('line')" +
            ".attr('x1',d=>data.nodes.find(n=>n.id===d.from).x)" +
            ".attr('y1',d=>data.nodes.find(n=>n.id===d.from).y)" +
            ".attr('x2',d=>data.nodes.find(n=>n.id===d.to).x)" +
            ".attr('y2',d=>data.nodes.find(n=>n.id===d.to).y)" +
            ".attr('class','link');" +
            "svg.selectAll('circle').data(data.nodes).enter().append('circle')" +
            ".attr('cx',d=>d.x).attr('cy',d=>d.y).attr('r',20)" +
            ".attr('fill',d=>d.color).attr('class','node');" +
            "svg.selectAll('text').data(data.nodes).enter().append('text')" +
            ".attr('x',d=>d.x).attr('y',d=>d.y+5)" +
            ".attr('text-anchor','middle').attr('class','label')" +
            ".text(d=>d.label);" +
            "</script></body></html>";
    }

    private String roleToColor(NodeRole role) {
        return switch (role) {
            case INFANT -> "#666666";
            case LEARNER -> "#0066ff";
            case ADULT -> "#00ff88";
            case SPECIALIST -> "#ffaa00";
            case GUARDIAN -> "#ff4444";
        };
    }

    public int getNodeCount() { return nodes.size(); }
    public int getEdgeCount() { return edges.size(); }
    public MapNode getNode(long id) { return nodes.get(id); }
}

File: docs-v2/assets/js/federation-map.js
/**
 * MATRIX Federation Map (W788)
 * Interactive D3.js-style force-directed graph using SVG.
 * NO LLM, NO server calls. Pure client-side.
 */

class FederationMap {
  constructor(containerId, width = 600, height = 400) {
    this.container = document.getElementById(containerId);
    if (!this.container) return;
    this.width = width;
    this.height = height;
    this.nodes = [];
    this.links = [];
    this.render();
  }

  addNode(id, role, x, y) {
    const colors = {
      'GUARDIAN': '#ff6b6b',
      'ADULT': '#4fc3f7',
      'LEARNER': '#51cf66',
      'INFANT': '#ffd43b'
    };
    this.nodes.push({ id, role, x, y, color: colors[role] || '#888' });
    this.render();
  }

  addLink(sourceId, targetId, strength = 1) {
    this.links.push({ source: sourceId, target: targetId, strength });
    this.render();
  }

  render() {
    let svg = `<svg width="${this.width}" height="${this.height}" xmlns="http://www.w3.org/2000/svg">`;

    // Links
    this.links.forEach(link => {
      const s = this.nodes.find(n => n.id === link.source);
      const t = this.nodes.find(n => n.id === link.target);
      if (s && t) {
        svg += `<line x1="${s.x}" y1="${s.y}" x2="${t.x}" y2="${t.y}" stroke="#4fc3f7" stroke-width="${link.strength}" opacity="0.4"/>`;
      }
    });

    // Nodes
    this.nodes.forEach(node => {
      svg += `<circle cx="${node.x}" cy="${node.y}" r="15" fill="${node.color}" opacity="0.8">
        <animate attributeName="r" values="15;20;15" dur="2s" repeatCount="indefinite"/>
      </circle>`;
      svg += `<text x="${node.x}" y="${node.y + 30}" text-anchor="middle" fill="white" font-size="10">${node.role}</text>`;
    });

    svg += '</svg>';
    this.container.innerHTML = svg;
  }
}

// Demo: Create a sample federation
document.addEventListener('DOMContentLoaded', () => {
  const map = new FederationMap('federation-map-container', 600, 400);
  if (!map.container) return;

  // Create sample nodes
  map.addNode(1, 'GUARDIAN', 300, 80);
  map.addNode(2, 'ADULT', 200, 200);
  map.addNode(3, 'ADULT', 400, 200);
  map.addNode(4, 'LEARNER', 150, 320);
  map.addNode(5, 'LEARNER', 350, 320);
  map.addNode(6, 'INFANT', 250, 380);
  map.addNode(7, 'INFANT', 450, 380);

  // Create links (mesh topology)
  map.addLink(1, 2, 2);
  map.addLink(1, 3, 2);
  map.addLink(2, 3, 1);
  map.addLink(2, 4, 1);
  map.addLink(3, 5, 1);
  map.addLink(4, 6, 1);
  map.addLink(5, 7, 1);
  map.addLink(2, 5, 1);
  map.addLink(3, 4, 1);
});
